File: src/big_config/git.clj
(ns big-config.git
  (:require [big-config.core :as core]
            [big-config.keys :as k]
            [big-config.run :as run]
            [big-config.utils :as u]))

(defn get-revision [revision key opts]
  (let [resolved (if (contains? opts revision) (get opts revision) revision)]
    (when-not (string? resolved)
      (throw (ex-info "Revision is neither a string nor a keyword"
                      {:revision revision :key key :opts opts})))
    (run/generic-cmd {:opts opts :cmd ["git" "rev-parse" resolved] :key key})))

(defn fetch-origin [opts]
  (run/generic-cmd {:opts opts :cmd ["git" "fetch" "origin"]}))

(defn upstream-name [key opts]
  (run/generic-cmd {:opts opts :cmd ["git" "rev-parse" "--abbrev-ref" "@{upstream}"] :key key}))

(defn git-diff [opts]
  (run/generic-cmd {:opts opts :cmd ["git" "diff" "--quiet"]}))

(defn git-push [opts]
  (run/generic-cmd {:opts opts :cmd ["git" "push"]}))

(defn compare-revisions [opts]
  (if (or (= (get opts k/git-prev-revision) (get opts k/git-origin-revision))
          (= (get opts k/git-current-revision) (get opts k/git-origin-revision)))
    (assoc opts k/exit 0 k/err nil)
    (assoc opts k/exit 1 k/err "The local revisions don't match the remote revision")))

(def check
  (core/create-workflow
   {:first-step :big-config.git/git-diff
    :wire-fn (fn [step]
               (case step
                 :big-config.git/git-diff [git-diff :big-config.git/fetch-origin]
                 :big-config.git/fetch-origin [fetch-origin :big-config.git/upstream-name]
                 :big-config.git/upstream-name [(fn [opts] (upstream-name k/git-upstream-name opts)) :big-config.git/pre-revision]
                 :big-config.git/pre-revision [(fn [opts] (get-revision "HEAD~1" k/git-prev-revision opts)) :big-config.git/current-revision]
                 :big-config.git/current-revision [(fn [opts] (get-revision "HEAD" k/git-current-revision opts)) :big-config.git/origin-revision]
                 :big-config.git/origin-revision [(fn [opts] (get-revision k/git-upstream-name k/git-origin-revision opts)) :big-config.git/compare-revisions]
                 :big-config.git/compare-revisions [compare-revisions :big-config.git/end]
                 :big-config.git/end [identity nil]
                 [identity nil]))}))

(u/register-function 'big-config.git/check check)
(u/register-function 'big-config.git/git-push git-push)

File: src/big_config/git.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from . import ERR, EXIT
from .core import workflow
from .run import generic_cmd

PREV_REVISION = "big-config.git/prev-revision"
CURRENT_REVISION = "big-config.git/current-revision"
ORIGIN_REVISION = "big-config.git/origin-revision"
UPSTREAM_NAME = "big-config.git/upstream-name"


def get_revision(revision: str, key: str, opts: Mapping[str, Any]) -> dict[str, Any]:
    rev = opts.get(revision, revision)
    if not isinstance(rev, str):
        raise ValueError(f"Revision is neither a string nor a keyword: {revision}")
    return generic_cmd(opts=opts, cmd=["git", "rev-parse", rev], key=key)


def fetch_origin(opts: Mapping[str, Any]) -> dict[str, Any]:
    return generic_cmd(opts=opts, cmd=["git", "fetch", "origin"])


def upstream_name(key: str, opts: Mapping[str, Any]) -> dict[str, Any]:
    return generic_cmd(opts=opts, cmd=["git", "rev-parse", "--abbrev-ref", "@{upstream}"], key=key)


def git_diff(opts: Mapping[str, Any]) -> dict[str, Any]:
    return generic_cmd(opts=opts, cmd=["git", "diff", "--quiet"])


def git_push(opts: Mapping[str, Any]) -> dict[str, Any]:
    return generic_cmd(opts=opts, cmd=["git", "push"])


def compare_revisions(opts: Mapping[str, Any]) -> dict[str, Any]:
    ok_revisions = opts.get(PREV_REVISION) == opts.get(ORIGIN_REVISION) or opts.get(CURRENT_REVISION) == opts.get(ORIGIN_REVISION)
    return {**opts, **({EXIT: 0, ERR: None} if ok_revisions else {EXIT: 1, ERR: "The local revisions don't match the remote revision"})}


def _wire(step: str, _step_fns: list[Any]):
    if step == "big-config.git/git-diff":
        return git_diff, "big-config.git/fetch-origin"
    if step == "big-config.git/fetch-origin":
        return fetch_origin, "big-config.git/upstream-name"
    if step == "big-config.git/upstream-name":
        return lambda opts: upstream_name(UPSTREAM_NAME, opts), "big-config.git/pre-revision"
    if step == "big-config.git/pre-revision":
        return lambda opts: get_revision("HEAD~1", PREV_REVISION, opts), "big-config.git/current-revision"
    if step == "big-config.git/current-revision":
        return lambda opts: get_revision("HEAD", CURRENT_REVISION, opts), "big-config.git/origin-revision"
    if step == "big-config.git/origin-revision":
        return lambda opts: get_revision(UPSTREAM_NAME, ORIGIN_REVISION, opts), "big-config.git/compare-revisions"
    if step == "big-config.git/compare-revisions":
        return compare_revisions, "big-config.git/end"
    return lambda x: x, None


check = workflow({"first_step": "big-config.git/git-diff", "wire_fn": _wire})
